### PR TITLE
feat: validate types when adding new tuples

### PR DESCRIPTION
**Pull request by miparnisari**
On every write, and if the authorization model has `schemaVersion == 1.1`, we will validate that the `user` part of the tuple satisfies the type definitions on the authorization model.

⚠️ Needs https://github.com/openfga/openfga/pull/247 to be merged so that if `schemaVersion == 1.1` and the auth model is incorrect, we throw an error.

## References
- https://github.com/openfga/rfcs/pull/7
- https://github.com/openfga/api/pull/27

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: server/commands/write.go
Comment (craigpastro) [inline server/commands/write.go:96]: Is this case still true?

Comment (miparnisari) [reply server/commands/write.go]: 😬 no

File: server/commands/write.go
Comment (craigpastro) [inline server/commands/write.go:112]: I am not sure, but I think you may have to worry if relations is nil? Also if the map contains `tk.Relation`. Perhaps something like 
```
relationInformation, ok := typeDefinition.GetMetadata().GetRelations()[tk.Relation]
```
may help. The `Gets` handle the nils.

Comment (miparnisari) [reply server/commands/write.go]: > The Gets handle the nils.

TIL! Thank you.

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:108]: Let's hold off on this PR a bit until the `typesystem` package gets merged in from the other PRs. A lot of this code can be reduced to helpers that will be in that package.

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:111]: We should be handling the error case where it's invalid differently, right?

For example, if we're using schema_version 1.0 then relations will be nil and we can return nil safely, but if schema_version is 1.1 and relations is nill, then that's a problem. Let's make sure we cover that non-nil error case here.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:28]: Refactor this to `tuple.NewTupleKey("repository:openfga/openfga", "administrator", "user:github|alice@openfga")`

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:47]: Shouldn't all of the prior tests still work without adding an explicit user type? I think we should be adding tests asserting new behavior for writes to models under the new schema_version 1.1. The existing tests are under schema_version 1.0 and should remain as is, yeah?

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:71]: Under the new `typesystem` package there should be a method called `IsDirectlyAssignable(objectType, relation string)` which will return 'true' if the relation is assignable and 'false' otherwise.

We should use that here instead.

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go]: You're right! When I wrote the original commit i didn't have the schema version available. 

Added it in recent commits and went with option b. The existing tests target 1.0, the ones i added target 1.1.

File: server/commands/write.go
Comment (miparnisari) [reply server/commands/write.go]: Cool, updated.

File: pkg/typesystem/typesystem.go
Comment (jon-whit) [inline pkg/typesystem/typesystem.go:255]: Can we please refactor this so that it's not associated with a rewrite but instead a relation itself.

e.g.

```
func (t *TypeSystem) IsDirectlyAssignable(relation *openfgapb.Relation) bool {
    rewrite := relation.GetUsersetRewrite()

    return containsSelf(rewrite)
}

func containsSelf(rewrite *openfgapb.Userset) bool {
    // copy/paste the current logic here
}
```
It's weird to me that we have an IsAssignable method that applies to a rewrite but not a Relation. Rewrites aren't assignable or not, relations are.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:769]: Let's omit these empty relation definitions.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:800]: Missing the typedef for `{Type: "user"}` here.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:838]: We're missing the type info for `group#member` here. Let's add it such that group#member is related to a user.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:860]: Can we please add a test case that makes sure the write with a user value such as `group:eng#member` is also invalid here. Since `reader` is restricted to `group`, a write such as `document:budget, reader, group:eng#member` is also invalid.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:868]: Can we also add a test case to make sure that Writes have to validate against any one of the directly related user types? All of the tests we have here only define 1 DirectlyRelatedUserTypes, and so it would be nice to add a test with 2 writes that succeeds because both of the writes validate against one of the two DirectlyRelatedUserTypes.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:994]: This isn't needed as we don't define any relations on it.

File: server/test/write.go
Comment (jon-whit) [inline server/test/write.go:1061]: I think it's probably better to just say "The wildcard '*' is an invalid user value for the 'document' relation 'reader'"

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:69]:
```suggestion
	var authModel *openfgapb.AuthorizationModel
```
for short

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:97]: Why do we need to pass `schemaVersion` if we're passing the `authModel`? The authModel contains the schema version. I vote we consolidate here.

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:120]: This is a static method. No need for it to be attached to an instance of `*WriteCommand`.

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:117]: (user=person:bob, relation=reader, object=doc:budget)

File: server/commands/write.go
Comment (jon-whit) [inline server/commands/write.go:128]: Please use the typesystem package here instead. You should be able to do:

```
relation, err := typesys.Relation(objectType, tk.GetRelation())
```
and then use the relation
```
relation.GetTypeInfo().GetDirectlyRelatedUserTypes()
// etc..
```

File: server/commands/write.go
Comment (craigpastro) [reply server/commands/write.go:117]: Or let's agree to use the `object#relation@user` notation from the paper everywhere to keep things shorter 😬

File: server/commands/write.go
Comment (craigpastro) [inline server/commands/write.go:141]: This error seems public facing so maybe:
```suggestion
             msg := "invalid authorization model"
			return serverErrors.NewInternalError(msg, errors.New(msg))
```

File: server/commands/write.go
Comment (craigpastro) [inline server/commands/write.go:139]: You could compare it to the enum if you would like.
```suggestion
		if schemaVersion == typesystem.SchemaVersion1_1 {
```

File: pkg/typesystem/typesystem.go
Comment (miparnisari) [reply pkg/typesystem/typesystem.go:255]: I can do that, but then I need to rollback my use of `typesystem.IsAssignable` in Write like you suggested, because I don't have the relation readily available there.

File: server/commands/write.go
Comment (miparnisari) [reply server/commands/write.go]: I added a `log.warn` in this method, so I need this :(

File: server/commands/write.go
Comment (miparnisari) [reply server/commands/write.go]: Haha @craigpastro i don't want to introduce a brand new notation :P

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go:1033]: 🤔 sorry I don't understand this

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go]: Hmmm i'm hesitant to say "wildcard". https://docs.fga.dev/modeling/basics/public-access#02-check-that-the-relationship-exists

File: server/test/write.go
Comment (jon-whit) [reply server/test/write.go:1033]: I just mean let's drop this, because it shouldn't be needed. If we omit the map then it will be nil, and that should be fine.

File: server/test/write.go
Comment (jon-whit) [reply server/test/write.go]: Why are you hesitant to say it? We refer to it in that doc page multiple times.
Screen Shot 2022-09-26 at 9 53 07 AM

File: server/commands/write.go
Comment (miparnisari) [reply server/commands/write.go]: I'm confused.. i don't want to return an error 🤔 I want to continue execution assuming a 1.0 model.

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go]: Because the doc says "it is not a wildcard" 😅

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go:1033]: Ahh okay. I was confused because I thought this should fail validation, but i realized i am not calling `commands.WriteAuthorizationModel` but `storage.WriteAuthorizationModel`

File: server/test/write.go
Comment (jon-whit) [reply server/test/write.go]: But the heading says "Wildcard Syntax Usage" 🤣

File: server/commands/write_test.go
Comment (craigpastro) [inline server/commands/write_test.go:134]: Is this actually used anywhere?

**Review by craigpastro — DISMISSED** [review]
LGTM

**Review by jon-whit — APPROVED** [review]

File: server/test/write.go
Comment (miparnisari) [reply server/test/write.go]: Following up this discussion: https://auth0.slack.com/archives/C033LQD4S11/p1664223016524409?thread_ts=1663955675.635259&cid=C033LQD4S11

**Pull request merged**
